subprojects/base-services/src/main/java/org/gradle/util/internal/GUtil.java

```diff
@@ -571,4 +573,27 @@ public static boolean isSecureUrl(URI url) {
         return !"http".equalsIgnoreCase(scheme);
     }
 
+    /**
+     * Checks the entry name for zip-slip vulnerable sequences.
+     *
+     * <b>IMPLEMENTATION NOTE</b>
+     * We do it this way instead of the way recommended in https://snyk.io/research/zip-slip-vulnerability
+     * for performance reasons, calling {@link File#getCanonicalPath()} is too expensive.
+     *
+     * @throws IllegalArgumentException if the entry contains vulnerable sequences
+     */
+    public static String safeZipEntryName(String name) {
+        if (isUnsafeZipEntryName(name)) {
+            throw new IllegalArgumentException(format("'%s' is not a safe zip entry name.", name));
+        }
+        return name;
+    }
+
+    public static boolean isUnsafeZipEntryName(String name) {
+        return name.isEmpty()
+            || name.startsWith("/")
+            || name.startsWith("\\")
+            || name.contains("..")
+            || name.contains(":");
```

Comment:
I wonder if the name containing `:` is too restrictive. I can imagine having file names which contain a `:`. 

I wonder if `FileUtils.normalize()` would help for the check here - we use this in some other places as a faster alternative to `canoncialize`.

Another alternative would also be to do the expensive check when the entry is unsafe. WDYT?

Reply:
We could disallow `:` only on Windows systems where the character is invalid anyway.

Reply:
I made the check for `:` Windows only.

The current implementation might prove to be too restrictive but it imposes the least amount of runtime overhead compared to other solutions so I think it's worth getting it through the RC phase.

Reply:
Should we go with https://github.com/gradle/gradle/pull/19592 ?